github/VisualStudio
# Create remote with the same name as upstream owner when viewing a PR

### The problem

When a user clones a forked repository and views a PR from upstream, we need to do a fetch on upstream in order to view files in the PR. Previously we were creating a temporary remote for upstream with a unique name, fetching from the remote and removing the remote afterwards.

Unfortunately git doesn't completely remove `config` file entries and users were ending up with *lots* of empty `[remote ...]` elements in their `.git/config` (one for every previewed PR)! 

For example:
```
[remote "jaredpar-8141ae63-26ec-4082-a75a-6cbe223bc357"]
[remote "jaredpar-3c0cb23a-068d-4386-9904-1a4ed48d2f96"]
[remote "jaredpar-9b39068d-6740-4bd3-8eff-2d6a1b01877f"]
```

### What this PR does

- If a remote with the required URL already exists, use that (i.e. use existing `upstream`)
- If no appropriate remote exists, create a new remote with the same name as the remote's owner
  - This is consistent with how we create a remote when checking out a PR
- If a remote with the name of the remote's owner already exists but with a different URL, fall back to using a temporary remote like we did before (e.g. if the existing remote uses `ssh://`)

### How to test

1. `git clone https://github.com/jcansdale/PullRequestSandbox`
2. `type .git/config`
```
[core]
        repositoryformatversion = 0
        filemode = false
        bare = false
        logallrefupdates = true
        symlinks = false
        ignorecase = true
[remote "origin"]
        url = https://github.com/jcansdale/PullRequestSandbox
        fetch = +refs/heads/*:refs/remotes/origin/*
[branch "master"]
        remote = origin
        merge = refs/heads/master
```
3. Open PR list from upstream in Visual Studio (this is the default list)
4. View a PR and open a file diff
5. `type .git/config`
```
[core]
        repositoryformatversion = 0
        filemode = false
        bare = false
        logallrefupdates = true
        symlinks = false
        ignorecase = true
[remote "origin"]
        url = https://github.com/jcansdale/PullRequestSandbox
        fetch = +refs/heads/*:refs/remotes/origin/*
[branch "master"]
        remote = origin
        merge = refs/heads/master
[remote "grokys"]
        url = https://github.com/grokys/PullRequestSandbox
        fetch = +refs/heads/*:refs/remotes/grokys/*
```
Should contain `[remote "grokys"]` with `url = https://github.com/grokys/PullRequestSandbox`.
6. Checkout the PR
7. `type .git/config`
```
[core]
        repositoryformatversion = 0
        filemode = false
        bare = false
        logallrefupdates = true
        symlinks = false
        ignorecase = true
[remote "origin"]
        url = https://github.com/jcansdale/PullRequestSandbox
        fetch = +refs/heads/*:refs/remotes/origin/*
[branch "master"]
        remote = origin
        merge = refs/heads/master
[remote "grokys"]
        url = https://github.com/grokys/PullRequestSandbox
        fetch = +refs/heads/*:refs/remotes/grokys/*
[branch "pr/54-testing-marking-the-branch-as-a-pr-4"]
        remote = grokys
        merge = refs/heads/testing-marking-4
        ghfvs-pr-owner-number = "grokys#54"
```
Should contain `[branch ...]` with `remote = grokys`.

### Related

- You can't fetch directly from a URL: https://github.com/libgit2/libgit2sharp/issues/1488
The `url` in the `Fetch(...)` method is actually a remote name not a URL. 😢  

Fixes #1749
Fixes #1753
